// @kentico/xperience-webpack-config/lib/kentico-xperience-webpack-config.js
const path = require('path');
const SystemJSPublicPathPlugin = require('systemjs-webpack-interop/SystemJSPublicPathWebpackPlugin');

module.exports = (opts) => {
    const argv = opts.argv || {};
    const isProduction = argv.p || argv.mode === 'production';

    return {
        mode: isProduction ? 'production' : 'development',
        entry: path.resolve(process.cwd(), 'src/entry.tsx'),
        output: {
            filename: isProduction ? 'entry.kxh.[contenthash].js' : 'entry.js',
            libraryTarget: 'system',
            path: path.resolve(process.cwd(), 'dist'),
            devtoolNamespace: `${opts.projectName}`,
            publicPath: ''
        },
        plugins: [
            new SystemJSPublicPathPlugin({
                systemjsModuleName: `@${opts.orgName}/${opts.projectName}`,
                rootDirectoryLevel: opts.rootDirectoryLevel
            })
        ],
        resolve: {
            extensions: ['.ts', '.tsx', '.js', '.jsx', '.json']
        },
        externals: [
            new RegExp('^@kentico/xperience-admin-'),
            'react',
            'react-dom',
            '@remix-run',
            'react-router',
            'react-router-dom',
            'react-beautiful-dnd'
        ]
    };
};
